packages/flutter/lib/src/material/time_picker.dart

```diff
@@ -1613,123 +1712,201 @@ class _TimePickerDialogState extends State<_TimePickerDialog> {
+      case TimePickerEntryMode.dial:
+        switch (orientation) {
+          case Orientation.portrait:
+            timePickerWidth = _kTimePickerWidthPortrait;
+            timePickerHeight = theme.materialTapTargetSize == MaterialTapTargetSize.padded
+                ? _kTimePickerHeightPortrait
+                : _kTimePickerHeightPortraitCollapsed;
+            break;
+          case Orientation.landscape:
+            timePickerWidth = _kTimePickerWidthLandscape * textScaleFactor;
+            timePickerHeight = theme.materialTapTargetSize == MaterialTapTargetSize.padded
+                ? _kTimePickerHeightLandscape
+                : _kTimePickerHeightLandscapeCollapsed;
+            break;
+        }
+        break;
+      case TimePickerEntryMode.input:
+        timePickerWidth = _kTimePickerWidthPortrait;
+        timePickerHeight = _kTimePickerHeightInput;
+        break;
+    }
+    return Size(timePickerWidth, timePickerHeight * textScaleFactor);
+  }
+
   @override
   Widget build(BuildContext context) {
     assert(debugCheckHasMediaQuery(context));
     final MediaQueryData media = MediaQuery.of(context);
     final TimeOfDayFormat timeOfDayFormat = localizations.timeOfDayFormat(alwaysUse24HourFormat: media.alwaysUse24HourFormat);
     final bool use24HourDials = hourFormat(of: timeOfDayFormat) != HourFormat.h;
     final ThemeData theme = Theme.of(context);
+    final ShapeBorder shape = TimePickerTheme.of(context).shape ?? _kDefaultShape;
+    final Orientation orientation = MediaQuery.of(context).orientation;
 
-    final Widget picker = Padding(
-      padding: const EdgeInsets.all(16.0),
-      child: AspectRatio(
-        aspectRatio: 1.0,
-        child: _Dial(
-          mode: _mode,
-          use24HourDials: use24HourDials,
-          selectedTime: _selectedTime,
-          onChanged: _handleTimeChanged,
-          onHourSelected: _handleHourSelected,
-        ),
-      ),
-    );
-
-    final Widget actions = ButtonBar(
+    Color toggleColor;
+    switch (theme.brightness) {
+      case Brightness.light:
+        toggleColor = Colors.grey[700];
+        break;
+      case Brightness.dark:
+        toggleColor = Colors.white;
+        break;
+    }
+    final Widget actions = Row(
       children: <Widget>[
-        FlatButton(
-          child: Text(localizations.cancelButtonLabel),
-          onPressed: _handleCancel,
+        const SizedBox(width: 10.0),
+        IconButton(
+          color: toggleColor,
+          onPressed: _handleEntryModeToggle,
+          icon: Icon(_entryMode == TimePickerEntryMode.dial ? Icons.keyboard : Icons.access_time),
+          // TODO(rami-a): Localize these strings.
+          tooltip: _entryMode == TimePickerEntryMode.dial
+              ? 'Switch to text input mode'
+              : 'Switch to dial picker mode',
         ),
-        FlatButton(
-          child: Text(localizations.okButtonLabel),
-          onPressed: _handleOk,
+        Expanded(
+          child: ButtonBar(
```

Comment:
ButtonBar is problematic. Can we avoid it here?

Reply:
I think ButtonBar gives some helpful logic (like moving to a Column layout if the buttons don't fit) so I'd like to keep it for now. I'll add a TODO to move away from this in the future.